### PR TITLE
Improve handling of multiple RHI::CommandList threads in BuildCommandList

### PR DESCRIPTION
**Pull request by dmcdiarmid-ly**
Added the submit range to the FrameGraphExecuteContext to make parallel command list handling more consistent across scopes and passes.

All AtomSampleViewer tests passed
Tested AtomSampleViewer DX12/Vulkan
Tested Editor DX12/Vulkan
Tested MaterialEditor DX12/Vulkan

### CODE REVIEW

File: Gems/Atom/RHI/DX12/Code/Source/RHI/CommandList.cpp [hunk 1]
```diff
@@ -404,6 +410,8 @@ namespace AZ
 
         void CommandList::Submit(const RHI::DrawItem& drawItem)
         {
+            ValidateSubmitItem(drawItem);
```
Comment (antonmic) [inline Gems/Atom/RHI/DX12/Code/Source/RHI/CommandList.cpp:413]: Just a note for the future, might be good to have a common function that all the backend implementations call where we can put common functionality like the submit item validation.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h [hunk 2]
```diff
@@ -76,7 +77,7 @@ namespace AZ
             /// Submits a single dispatch item for processing on the command list.
             virtual void Submit(const DispatchItem& dispatchItem) = 0;
 
-            //! Submits a single dispatch rays item for processing on the command list.
+            /// Submits a single dispatch rays item for processing on the command list.
```
Comment (rgba16f) [inline Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h:80]: I believe //! is the current standard for Doxygen in o3de.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h [hunk 3]
```diff
@@ -85,11 +86,42 @@ namespace AZ
...
+            {
+                // the zero-based start index of the range
+                uint32_t m_startIndex = 0;
+
+                // the end index of the range
+                // Note: this is an exclusive index, meaning submitted item indices should be less than this index
+                uint32_t m_endIndex = 0;
+
+                // returns the number of items in the range
+                uint32_t GetCount() const { return m_endIndex - m_startIndex; }
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h:107]: add 'inline' keyword

File: Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h [hunk 4]
```diff
@@ -85,11 +86,42 @@ namespace AZ
...
+            void SetSubmitRange(const SubmitRange& submitRange) { m_submitRange = submitRange; }
+
+            /// Validates a submit item against the range for this command list, and tracks the total number of submits
+            void ValidateSubmitItem(const SubmitItem& submitItem);
+
+            /// Validates the total number of submits against the expected number
+            void ValidateTotalSubmits(const ScopeProducer* scopeProducer);
+
+            /// Resets the total number of submits
+            void ResetTotalSubmits() { m_totalSubmits = 0; }
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h:120]: inline keyword

File: Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteContext.h [hunk 5]
```diff
@@ -48,6 +49,9 @@ namespace AZ
 
             /// Returns the command list associated with the index in the batch.
             CommandList* GetCommandList() const;
+
+            /// Returns the submit range for this context.
+            const CommandList::SubmitRange& GetSubmitRange() const { return m_descriptor.m_submitRange; }
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteContext.h:54]: It is a bit weird to see a bunch of getters and setters and only this one implemented in the headers. Lets move the implementation to the cpp to match the other functions here.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h [hunk 6]
```diff
@@ -85,11 +86,42 @@ namespace AZ
...
+            struct SubmitRange
+            {
+                // the zero-based start index of the range
+                uint32_t m_startIndex = 0;
+
+                // the end index of the range
+                // Note: this is an exclusive index, meaning submitted item indices should be less than this index
+                uint32_t m_endIndex = 0;
+
+                // returns the number of items in the range
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h:106]: //!

File: Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteGroup.h [hunk 7]
```diff
@@ -98,10 +101,16 @@ namespace AZ
                 /// user wishes to fill in the command list at context creation time.
                 CommandList* m_commandList = nullptr;
 
-                /// An ordered list of scope ids in the group.
-                const ScopeId* m_scopeIds = nullptr;
+                /// An ordered list of scope ids and submit counts in the group.
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteGroup.h:104]: the ScopeEntry struct does not contain a list. Is this a stale comment?

File: Gems/Atom/RHI/Code/Include/Atom/RHI/SubmitItem.h [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+#pragma once
+
+namespace AZ
+{
+    namespace RHI
+    {
+        /// Base for the command list Submit item types.
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Include/Atom/RHI/SubmitItem.h:14]: Minor: I understand where for existing classes you are following the existing commenting convention but for newer classes lets follow the correct commenting standard which is //! for class/api definitions and // for everything else

File: Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuteGroupMerged.cpp [hunk 9]
```diff
@@ -29,11 +29,11 @@ namespace AZ
             m_workRequest.m_signalFence = m_scopes.back()->GetSignalFenceValue();
             m_workRequest.m_commandLists.resize(1);
 
-            AZStd::vector<RHI::ScopeId> scopeIds;
-            scopeIds.reserve(m_scopes.size());
+            AZStd::vector<InitMergedRequest::ScopeEntry> scopeEntries;
```
Comment (moudgils) [inline Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuteGroupMerged.cpp:32]: you will need to update the metal version as well. It should be similar to dx12

File: Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuteGroup.cpp [hunk 10]
```diff
@@ -48,6 +48,7 @@ namespace AZ
 
             InitRequest request;
             request.m_scopeId = scope.GetId();
+            request.m_submitCount = scope.GetEstimatedItemCount();
```
Comment (moudgils) [inline Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuteGroup.cpp:51]: you will need to update the metal version as well. It should be similar to dx12

File: Gems/Atom/RHI/Code/Source/RHI/FrameGraphExecuteGroup.cpp [hunk 11]
```diff
@@ -41,10 +42,13 @@ namespace AZ
             descriptor.m_scopeId = request.m_scopeId;
             descriptor.m_commandListCount = request.m_commandListCount;
 
+            uint32_t submitCount = request.m_submitCount;
+            uint32_t commandListCount = request.m_commandListCount;
             for (uint32_t i = 0; i < request.m_commandListCount; ++i)
             {
                 descriptor.m_commandListIndex = i;
                 descriptor.m_commandList = request.m_commandLists[i];
+                descriptor.m_submitRange = { (i * submitCount) / commandListCount, ((i + 1) * submitCount) / commandListCount };
```
Comment (moudgils) [inline Gems/Atom/RHI/Code/Source/RHI/FrameGraphExecuteGroup.cpp:51]: Can you add a comment explaining the submit range per command list. Basically explaining how each command list is assigned to do 'submitCount/commandListCount' worth of work.

File: Gems/Atom/RPI/Code/Source/RPI.Public/Pass/ComputePass.cpp [hunk 12]
```diff
@@ -140,7 +140,6 @@ namespace AZ
         void ComputePass::SetupFrameGraphDependencies(RHI::FrameGraphInterface frameGraph)
         {
             RenderPass::SetupFrameGraphDependencies(frameGraph);
-            frameGraph.SetEstimatedItemCount(1);
```
Comment (moudgils) [inline Gems/Atom/RPI/Code/Source/RPI.Public/Pass/ComputePass.cpp:143]: why was this removed?

File: Gems/AtomTressFX/Code/Passes/HairSkinningComputePass.cpp [hunk 13]
```diff
@@ -225,8 +232,13 @@ namespace AZ
                 // This includes the PerView, PerScene and PerPass srgs (what about per draw?)
                 SetSrgsForDispatch(commandList);
 
-                for (const RHI::DispatchItem* dispatchItem : m_dispatchItems)
+                AZStd::unordered_set<const RHI::DispatchItem*>::iterator it = m_dispatchItems.begin();
+                AZStd::advance(it, context.GetSubmitRange().m_startIndex);
+
+                for (uint32_t index = context.GetSubmitRange().m_startIndex; index < context.GetSubmitRange().m_endIndex; ++index, ++it)
```
Comment (moudgils) [inline Gems/AtomTressFX/Code/Passes/HairSkinningComputePass.cpp:238]: A quick comment to say that we are submitting items across the range provided by RHI

File: Gems/Atom/Feature/Common/Code/Source/DiffuseGlobalIllumination/DiffuseProbeGridBorderUpdatePass.cpp [hunk 14]
```diff
@@ -119,6 +119,9 @@ namespace AZ
 
             RPI::Scene* scene = m_pipeline->GetScene();
             DiffuseProbeGridFeatureProcessor* diffuseProbeGridFeatureProcessor = scene->GetFeatureProcessor<DiffuseProbeGridFeatureProcessor>();
+
+            frameGraph.SetEstimatedItemCount(aznumeric_cast<uint32_t>(diffuseProbeGridFeatureProcessor->GetVisibleRealTimeProbeGrids().size() * 4));
```
Comment (moudgils) [inline Gems/Atom/Feature/Common/Code/Source/DiffuseGlobalIllumination/DiffuseProbeGridBorderUpdatePass.cpp:123]: Add a comment to explain the *4

**Review by moudgils — COMMENTED** [review]
Nice work. Minor comments.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h (hunk 3 above)
Comment (dmcdiarmid-ly) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h:107]: The inline keyword is not necessary if the method implementation is in the class definition, it is automatically inlined.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h (hunk 4 above)
Comment (dmcdiarmid-ly) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h:120]: Same as above.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h (hunk 6 above)
Comment (dmcdiarmid-ly) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h:106]: Are //! (or ///) required in the private (non-interface) portions of a class? I thought it was only for interface documentation comments.

Comment (moudgils) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/CommandList.h:106]: We can leave this as is as you are following the pattern that already exists in this class. I usually try to convert the style for a file if I am updating it in any way but it is fine if we dont do that here. I do eventually plan on updating is all in one go if I can carve out some time for it.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteContext.h (hunk 5 above)
Comment (moudgils) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteContext.h:54]: This adds more cruft to the code. It would be nice if you can move the getter/setters from cpp to the header than to be consistent. Thanks.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteGroup.h (hunk 7 above)
Comment (moudgils) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteGroup.h:104]: Ahh ok. Maybe remove the extra line between ScopeEntry and m_scopeEntries just to make it obvious.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteContext.h (hunk 5 above)
Comment (dmcdiarmid-ly) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteContext.h:54]: I moved the rest of the accessors to the header.  Honestly it seems like we should just provide a single accessor for the Descriptor at this point, but that's an exercise for another PR.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteGroup.h (hunk 7 above)
Comment (dmcdiarmid-ly) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/FrameGraphExecuteGroup.h:104]: Comment moved.

File: Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuteGroup.cpp (hunk 10 above)
Comment (dmcdiarmid-ly) [reply Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuteGroup.cpp:51]: Updated.

File: Gems/Atom/RHI/Code/Source/RHI/FrameGraphExecuteGroup.cpp (hunk 11 above)
Comment (dmcdiarmid-ly) [reply Gems/Atom/RHI/Code/Source/RHI/FrameGraphExecuteGroup.cpp:51]: Added comment.

File: Gems/Atom/Feature/Common/Code/Source/DiffuseGlobalIllumination/DiffuseProbeGridBorderUpdatePass.cpp (hunk 14 above)
Comment (dmcdiarmid-ly) [reply Gems/Atom/Feature/Common/Code/Source/DiffuseGlobalIllumination/DiffuseProbeGridBorderUpdatePass.cpp:123]: Added comment.

File: Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuteGroupMerged.cpp (hunk 9 above)
Comment (dmcdiarmid-ly) [reply Gems/Atom/RHI/DX12/Code/Source/RHI/FrameGraphExecuteGroupMerged.cpp:32]: Updated.

**Review by moudgils — APPROVED** [review]
